### ISSUE TITLE
V8 : Back office  freeze when opening multiple content nodes

### ISSUE THREAD

**Issue by stherm** [umbraco/Umbraco-CMS#5484]
When opening multiple node in a tree view in the backOffice At some point the back office stop. 

Reproduction
I have Contents that are kind of a tree, so multiples root,multiple childs and the child have multiple childs and so on. 

If I start to open all childs, at some point the back office will just stop rendering and freeze. The only way to resolve it is by refreshing the page. 

I've teste with SQLCE and SQL Server , happen with both. 

I get the bug between 26 and 40 nodes open.

![image](https://user-images.githubusercontent.com/1386693/57952373-fc151100-78ba-11e9-9b11-21886717a433.png)

**Comment by stherm** [comment]
For more information, here is the javascript profiler 
![image](https://user-images.githubusercontent.com/1386693/58100130-08df8080-7bab-11e9-80d2-83b994d9000a.png)

I've check and the problem doesn't weems to be DB related

**Comment by nul800sebastiaan** [comment]
@stherm I am not sure what you mean with the nodes being "open"? Do you mean that when you expand 26+ tree items the backoffice stops responding in your browser?

**Comment by stherm** [comment]
Basically yes. I'm just expanding the tree and at some points, the browser freeze. It's not always at the same number of node/document. but I eventually get it to completely freeze in under 50 nodes

**Comment by stherm** [comment]
I have some more details on the bug : 
There's some tests I've done 

All node where name 1,2,3,4, 1.1, 1.2 , 1.3 for test purpose

Not linked with tree depth - 27 levels ok (Although I can't create a 28th node due to sql error)

Test 2

Banch 1->27 node | Branch 2->27 node (pretty slow to render)

Banch 1->27 nodes | Branch 2->27 nodes  | branch 3-> 27 node (really slow to render)

Branche 4-> 27 node  | Branch 2 (flat structure) -> 15 nodes | branch 3 -> 4 nodes Branche 5 -> 1/27 node (FAIL and Freeze)

**Comment by emmaburstow** [comment]
Hi everyone,

I can't recreate this on my latest install. I've created 5 parent nodes, each with 28 child nodes successfully, all open at the same time without freezing. 

![issue5484](https://user-images.githubusercontent.com/37150989/58865806-22a2bc80-86af-11e9-850e-352b9253579c.gif)

Emma

**Comment by stherm** [comment]
Hi I've take a screen shot of the two bug I've descibe before. The first one is an SQL error related to the length of the url. That one I get. 
![téléchargement (1)](https://user-images.githubusercontent.com/1386693/58882885-5380f280-86ab-11e9-8aa3-2508726c7fe1.png)

The second one, it's the one where the back office freeze. Lie I've said, with a structure of parent - > child -> subchild this happend. Doesn't have to be has extreme has the screen shot, but if you have multiple nodes that are in that kind of structure, it will happend. 

![téléchargement](https://user-images.githubusercontent.com/1386693/58882891-5845a680-86ab-11e9-8584-cec0e5a0d99b.png)

**Comment by nul800sebastiaan** [comment]
@stherm It looks like you have a test site already, with a SQL CE database, can you zip it up and upload it somewhere so we can have a look so we don't have to try and replicate the problems? 

Make sure to change the admin username/password beforehand and send it along as well so we can log in easily.

**Comment by stherm** [comment]
Hi, I,ve make a test site which replicate the bug. Just try and open all node, it whatever order and it'll freeze at one point. Also, it's impossible to create a 29th node in the first structure (which I know is ridiculous, but it's a bug :) ). In my case, for this site, i've opened all node of root3 and root2 and it strarted to freeze opening nodes in root1.

The zip is too big to share here : [https://drive.google.com/drive/folders/1CEnRyjKXSuZk-MX2kOBbSe5rIHmdLgLX?usp=sharing]

**Comment by danimalx** [comment]
I can confirm the same behaviour. First it slows down to a crawl, then you either sit there and wait for the tab to crash or end it manually. It's happened to lots of my users and is very frustrating for the front end editor.

**Comment by bdsoko** [comment]
I can confirm this issue as well.  We are running v8.1.1.

**Comment by bdsoko** [comment]
Digging a little deeper, I believe the issue has to do with the _trackExpandedPaths  method in the umbraco.services.js.
Not entirely sure what the purpose of this method is; however, while iterating through the expandedPaths array, duplicate paths are added to the array.
For example, navigating 7 levels deep causes several duplicates in the expandedPaths array.  In this instance, 71 items in the array (see below).  Now imagine the size of this array as you continue to navigate the tree opening nodes.  My test causes Umbraco to crash after opening 20+ nodes.  I do have a working fix for this, just don’t know how-to submit it.
Generated using console.log(expandedPaths); right after the _.each(clonedPaths, function (p) {} call.
0: "-1,-20"
1: "-1,1565,2063"
2: "-1,1565,1858,1977"
3: "-1,1565,1858,1977"
4: "-1,1565,1858,1923,2125"
5: "-1,1565,1858,1923,2125"
6: "-1,1565,1858,1923,2125"
7: "-1,1565,1858,1923,2125"
8: "-1,1565,1858,1923,1925,1927,1931"
9: "-1,1565,1858,1923,1925,1927,1931"
10: "-1,1565,1858,1923,1925,1927,1931"
11: "-1,1565,1858,1923,1925,1927,1931"
12: "-1,1565,1858,1923,1925,1927,1931"
13: "-1,1565,1858,1923,1925,1927,1931"
14: "-1,1565,1858,1923,1925,1927,1931"
15: "-1,1565,1858,1923,1925,1927,1931"
16: "-1,1565,1858,1923,1925,1927,1931"
17: "-1,1565,1858,1923,1925,1927,1931"
18: "-1,1565,1858,1923,1925,1927,1931"
19: "-1,1565,1858,1923,1925,1927,1931"
20: "-1,1565,1858,1923,1925,1927,1931"
21: "-1,1565,1858,1923,1925,1927,1931"
22: "-1,1565,1858,1923,1925,1927,1931"
23: "-1,1565,1858,1923,1925,1927,1931"
24: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
25: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
26: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
27: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
28: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
29: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
30: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
31: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
32: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
33: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
34: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
35: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
36: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
37: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
38: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
39: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
40: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
41: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
42: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
43: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
44: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
45: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
46: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
47: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
48: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
49: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
50: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
51: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
52: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
53: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
54: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
55: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
56: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
57: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
58: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
59: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
60: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
61: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
62: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
63: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
64: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
65: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
66: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
67: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
68: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
69: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
70: "-1,1565,1858,1923,1925,1927,1929,1942,1943"
71: "-1,1565,1858,1923,1925,1927,1929,1942,1943"

**Comment by stherm** [comment]
@bdsoko Have you tested this with a large number of nodes and subnodes? Could you send me the code or a test site with your fix? I would do some test with it.

**Comment by bdsoko** [comment]
@stherm I downloaded your test project and ran a test pre and post fix.  Before applying the fix, I opened all nodes under 'Root', then opened 'Root2'.  In the screenshot below, you'll notice that the number of items in the array jumps from 29 to 58, just by opening the 'Root2' node.  You'll also notice a lot of duplicates.  After applying my fix, after opening both 'Root' and 'Root2', the number of items in the array is only 4.  I'm no expert on Umbraco, but I think issue has do with that the code doesn't check if the item is already in the array while looping.  FYI, I also opened all your nodes in your test project without error, post fix.

BEFORE FIX:
![before fix](https://user-images.githubusercontent.com/11299552/62384165-3fc10380-b517-11e9-9654-f894790a59fb.png)

AFTER FIX:
![after fix](https://user-images.githubusercontent.com/11299552/62384183-4c455c00-b517-11e9-8421-c4ca89442963.png)
